broker/artemis/plugin/src/main/java/org/eclipse/kapua/broker/artemis/plugin/security/context/SecurityContextHandler.java

```diff
@@ -12,127 +12,262 @@
+            appendDetailedServerContextReport(builder, caller, connectionId);
+            break;
+        case Compact:
+            appendServerContextReport(builder, server);
+            break;
+        case DetailedServer:
+            appendDetailedServerContextReport(builder, caller, connectionId);
+            break;
+        default:
+            break;
+        }
+        builder.append(REPORT_HEADER);
+        logger.info("{}", builder);
+    }
+
+    public void init(ActiveMQServer server) {
+       executorWrapper = new ExecutorWrapper("ServerReport", () -> printReport(server, "ServerReportTask", "N/A"), 60, 30, TimeUnit.SECONDS);
+       executorWrapper.start();
     }
 
-    public void setSessionContext(SessionContext sessionContext) {
-        principalMap.put(sessionContext.getConnectionId(), sessionContext.getPrincipal());
-        sessionContextMapByConnection.put(sessionContext.getConnectionId(), sessionContext);
+    public void shutdown(ActiveMQServer server) {
+        if (executorWrapper!=null) {
+            executorWrapper.stop();
+        }
     }
 
-    public void setSessionContext(SessionContext sessionContext, AuthResponse authResponse) throws KapuaIllegalArgumentException {
-        setSessionContext(sessionContext);
-        aclMap.put(sessionContext.getConnectionId(), new Acl(sessionContext.getPrincipal(), authResponse));
-        sessionContextMapByClient.put(Utils.getFullClientId(sessionContext), sessionContext);
+    public ConnectionToken updateConnectionTokenOnConnection(String connectionId) {
+        ConnectionToken connectionToken = connectionTokenCache.getAndRemove(connectionId);
+        if (connectionToken==null) {
+            connectionTokenCache.put(connectionId,
+                new ConnectionToken(SecurityAction.brokerConnect, KapuaDateUtils.getKapuaSysDate()));
+        }
+        else {
+            //the disconnect callback is called after the connect so nothing to add to the context
+            //TODO add metric?
+            logger.warn("Connect callback called after the disconnection callback ({} - {} - {})", connectionId, connectionToken.getAction(), connectionToken.getActionDate());
+        }
+        return connectionToken;
     }
 
-    private SessionContext cleanSessionContext(SessionContext sessionContext) {
-        principalMap.remove(sessionContext.getConnectionId());
-        sessionContextMapByConnection.remove(sessionContext.getConnectionId());
-        if (!sessionContext.isInternal()) {
-            aclMap.remove(sessionContext.getConnectionId());
-            //check for stealing link
-            //remove object only if connection ids match
+    public boolean setSessionContext(SessionContext sessionContext, List<AuthAcl> authAcls) throws KapuaIllegalArgumentException {
+        logger.info("Updating session context for connection id: {}", sessionContext.getConnectionId());
+        synchronized (sessionContext.getConnectionId().intern()) {
```

Comment:
good point, I'm investigating possible solutions. I don't want to implement a global lock since it becomes a performance bottleneck.
For that reason I supposed the intern() was the best choice but it doesn't

Reply:
I pushed  fix using a Lock array (to try to parallelize the processing).
I think it's good sicne we can tune the size of the array according to the threading model.
The lock is selected using the connectionId hash so each connection operation will use the same lock.
I'm still testing the solution and I'm looking forward for your review!

Reply:
I looked at your commit, and I think it should work, though more complicated than I would have implemented since it adds an additional class, and it's more complicated to understand the logic when reading the code, whereas seeing the `synchronized(connectionId)` it's more clear to me what it's doing.

Combining what I had in mind with your solution, you could do something like:
```java
    private static final Object[] LOCKS = new Object[LOCKS_SIZE];
    // then initialize each item with "new Object()";

    // Choose a lock based on the connectionId (or any string)
    private static Object getLock(String connectionId) {
        return LOCKS[Math.abs(connectionId.hashCode()%LOCKS_SIZE)];
    }

    // To use the lock
    yourMethod() {
        synchronized(getLock(connectionId)) {
            // Run code
        }
    }
```

It's fine for me either way; I think both should be safer than the previous intern() method.

Reply:
it's not too clear to me the solution you proposed. I suggest to have a call to discuss about it.
The class I added was just to encapsulate the code should be run inside the synchronized block (through synchonized or lock the meaning is the same)

Reply:
I was looking again into your solution and it's getting more clear to me. Yes, it seems more or less what I implemented with the proper translation (assuming the class I added was just to encapsulate code should be ran inside the synchronized block so it's not part of the core implementation):
```
Object -> Lock
synchonized -> lock.lock() .... lock.unlock()
```
I prefer to stay with my implementation but it's better to discuss in a call, as I suggest in my previous post, if you agree.